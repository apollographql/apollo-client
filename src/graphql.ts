/**
 * This type represents the dynamic form of any data that may be returned from
 * a GraphQL API. Basically this is just a narrower `mixed` type, or another
 * type for JSON.
 */
export type GraphQLData = null | boolean | number | string | GraphQLArrayData | GraphQLObjectData;

/**
 * An that may be returned from a GraphQL API.
 */
export interface GraphQLObjectData { [fieldName: string]: GraphQLData; }

/**
 * A list of values that may be returned from a GraphQL API.
 */
export interface GraphQLArrayData extends Array<GraphQLData> {}

/**
 * A guard function which checks to see if the provided value is GraphQL object
 * data. True if it is.
 */
export function isObjectData (value: any): value is GraphQLObjectData {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}
