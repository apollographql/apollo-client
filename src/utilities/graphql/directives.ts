// Provides the methods that allow QueryManager to handle the `skip` and
// `include` directives within GraphQL.
import {
  SelectionNode,
  VariableNode,
  BooleanValueNode,
  DirectiveNode,
  DocumentNode,
  ArgumentNode,
  ValueNode,
  ASTNode,
} from 'graphql';

import { visit } from 'graphql/language/visitor';

import { invariant } from 'ts-invariant';

export type DirectiveInfo = {
  [fieldName: string]: { [argName: string]: any };
};

export function shouldInclude(
  { directives }: SelectionNode,
  variables?: Record<string, any>,
): boolean {
  if (!directives || !directives.length) {
    return true;
  }
  return getInclusionDirectives(
    directives
  ).every(({ directive, ifArgument }) => {
    let evaledValue: boolean = false;
    if (ifArgument.value.kind === 'Variable') {
      evaledValue = variables && variables[(ifArgument.value as VariableNode).name.value];
      invariant(
        evaledValue !== void 0,
        `Invalid variable referenced in @${directive.name.value} directive.`,
      );
    } else {
      evaledValue = (ifArgument.value as BooleanValueNode).value;
    }
    return directive.name.value === 'skip' ? !evaledValue : evaledValue;
  });
}

export function getDirectiveNames(root: ASTNode) {
  const names: string[] = [];

  visit(root, {
    Directive(node: DirectiveNode) {
      names.push(node.name.value);
    },
  });

  return names;
}

export function hasDirectives(names: string[], root: ASTNode) {
  return getDirectiveNames(root).some(
    (name: string) => names.indexOf(name) > -1,
  );
}

export function hasClientExports(document: DocumentNode) {
  return (
    document &&
    hasDirectives(['client'], document) &&
    hasDirectives(['export'], document)
  );
}

export type InclusionDirectives = Array<{
  directive: DirectiveNode;
  ifArgument: ArgumentNode;
}>;

function isInclusionDirective({ name: { value } }: DirectiveNode): boolean {
  return value === 'skip' || value === 'include';
}

export function getInclusionDirectives(
  directives: ReadonlyArray<DirectiveNode>,
): InclusionDirectives {
  const result: InclusionDirectives = [];

  if (directives && directives.length) {
    directives.forEach(directive => {
      if (!isInclusionDirective(directive)) return;

      const directiveArguments = directive.arguments;
      const directiveName = directive.name.value;

      invariant(
        directiveArguments && directiveArguments.length === 1,
        `Incorrect number of arguments for the @${directiveName} directive.`,
      );

      const ifArgument = directiveArguments![0];
      invariant(
        ifArgument.name && ifArgument.name.value === 'if',
        `Invalid argument for the @${directiveName} directive.`,
      );

      const ifValue: ValueNode = ifArgument.value;

      // means it has to be a variable value if this is a valid @skip or @include directive
      invariant(
        ifValue &&
          (ifValue.kind === 'Variable' || ifValue.kind === 'BooleanValue'),
        `Argument for the @${directiveName} directive must be a variable or a boolean value.`,
      );

      result.push({ directive, ifArgument });
    });
  }

  return result;
}

