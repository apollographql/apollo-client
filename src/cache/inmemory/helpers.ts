import { SelectionSetNode } from 'graphql';

import { NormalizedCache } from './types';
import {
  Reference,
  isReference,
  StoreValue,
  StoreObject,
  isField,
  DeepMerger,
  resultKeyNameFromField,
  shouldInclude,
} from '../../utilities';

export const hasOwn = Object.prototype.hasOwnProperty;

export function getTypenameFromStoreObject(
  store: NormalizedCache,
  objectOrReference: StoreObject | Reference,
): string | undefined {
  return isReference(objectOrReference)
    ? store.get(objectOrReference.__ref, "__typename") as string
    : objectOrReference && objectOrReference.__typename;
}

export const TypeOrFieldNameRegExp = /^[_a-z][_0-9a-z]*/i;

export function fieldNameFromStoreName(storeFieldName: string): string {
  const match = storeFieldName.match(TypeOrFieldNameRegExp);
  return match ? match[0] : storeFieldName;
}

export function selectionSetMatchesResult(
  selectionSet: SelectionSetNode,
  result: Record<string, any>,
  variables?: Record<string, any>,
): boolean {
  if (result && typeof result === "object") {
    return Array.isArray(result)
      ? result.every(item => selectionSetMatchesResult(selectionSet, item, variables))
      : selectionSet.selections.every(field => {
        if (isField(field) && shouldInclude(field, variables)) {
          const key = resultKeyNameFromField(field);
          return hasOwn.call(result, key) &&
            (!field.selectionSet ||
             selectionSetMatchesResult(field.selectionSet, result[key], variables));
        }
        // If the selection has been skipped with @skip(true) or
        // @include(false), it should not count against the matching. If
        // the selection is not a field, it must be a fragment (inline or
        // named). We will determine if selectionSetMatchesResult for that
        // fragment when we get to it, so for now we return true.
        return true;
      });
  }
  return false;
}

export function storeValueIsStoreObject(
  value: StoreValue,
): value is StoreObject {
  return value !== null &&
    typeof value === "object" &&
    !isReference(value) &&
    !Array.isArray(value);
}

export function makeProcessedFieldsMerger() {
  return new DeepMerger;
}
