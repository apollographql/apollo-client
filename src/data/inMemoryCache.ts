import { Cache, CacheWrite } from './cache';

import { DocumentNode } from 'graphql';

import { NormalizedCache } from './storeUtils';

import { ApolloReducerConfig } from '../store';

import { writeResultToStore } from './writeToStore';

import {
  readQueryFromStore,
  diffQueryAgainstStore,
  DiffResult,
} from './readFromStore';

export type OptimisticStoreItem = {
  id: string;
  data: NormalizedCache;
  transaction: (c: Cache) => void;
};

export class InMemoryCache extends Cache {
  private data: NormalizedCache;
  private config: ApolloReducerConfig;
  private optimistic: OptimisticStoreItem[] = [];

  constructor(config: ApolloReducerConfig, initialStore: NormalizedCache = {}) {
    super(config.addTypename || false);
    this.config = config;
    this.data = initialStore;
  }

  public getData(): NormalizedCache {
    return this.data;
  }

  public getOptimisticData(): NormalizedCache {
    if (this.optimistic.length === 0) {
      return this.data;
    }

    const patches = this.optimistic.map(opt => opt.data);
    return Object.assign({}, this.data, ...patches) as NormalizedCache;
  }

  public getOptimisticQueue(): OptimisticStoreItem[] {
    return this.optimistic;
  }

  public reset(): Promise<void> {
    this.data = {};

    return Promise.resolve();
  }

  public applyTransformer(
    transform: (i: NormalizedCache) => NormalizedCache,
  ): void {
    this.data = transform(this.data);
  }

  public diffQuery(query: {
    query: DocumentNode;
    variables: any;
    returnPartialData?: boolean;
    previousResult?: any;
    optimistic: boolean;
  }): DiffResult {
    return diffQueryAgainstStore({
      store: query.optimistic ? this.getOptimisticData() : this.data,
      query: query.query,
      variables: query.variables,
      returnPartialData: query.returnPartialData,
      previousResult: query.previousResult,
      fragmentMatcherFunction: this.config.fragmentMatcher,
      config: this.config,
    });
  }

  public read(query: {
    query: DocumentNode;
    variables: any;
    rootId?: string;
    previousResult?: any;
    optimistic: boolean;
  }): any {
    if (query.rootId && typeof this.data[query.rootId] === 'undefined') {
      return null;
    }

    return readQueryFromStore({
      store: query.optimistic ? this.getOptimisticData() : this.data,
      query: query.query,
      variables: query.variables,
      rootId: query.rootId,
      fragmentMatcherFunction: this.config.fragmentMatcher,
      previousResult: query.previousResult,
      config: this.config,
    });
  }

  public writeResult(write: CacheWrite): void {
    writeResultToStore({
      ...write,
      store: this.data,
      dataIdFromObject: this.config.dataIdFromObject,
      fragmentMatcherFunction: this.config.fragmentMatcher,
    });
  }

  public removeOptimistic(id: string) {
    // Throw away optimistic changes of that particular mutation
    const toPerform = this.optimistic.filter(item => item.id !== id);

    this.optimistic = [];

    // Re-run all of our optimistic data actions on top of one another.
    toPerform.forEach(change => {
      this.recordOptimisticTransaction(change.transaction, change.id);
    });
  }

  public performTransaction(transaction: (c: Cache) => void) {
    // TODO: does this need to be different, or is this okay for an in-memory cache?
    transaction(this);
  }

  public recordOptimisticTransaction(
    transaction: (c: Cache) => void,
    id: string,
  ) {
    const before = this.getOptimisticData();

    const orig = this.data;
    this.data = { ...before };
    transaction(this);
    const after = this.data;
    this.data = orig;

    const patch: any = {};

    Object.keys(after).forEach(key => {
      if (after[key] !== before[key]) {
        patch[key] = after[key];
      }
    });

    this.optimistic.push({
      id,
      transaction,
      data: patch,
    });
  }
}
