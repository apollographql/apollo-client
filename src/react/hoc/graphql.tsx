import { DocumentNode } from 'graphql';

import { parser, DocumentType } from '../parser';
import { withQuery } from './query-hoc';
import { withMutation } from './mutation-hoc';
import { withSubscription } from './subscription-hoc';
import { OperationOption, DataProps, MutateProps } from './types';

export function graphql<
  TProps extends TGraphQLVariables | {} = {},
  TData = {},
  TGraphQLVariables = {},
  TChildProps = Partial<DataProps<TData, TGraphQLVariables>> &
    Partial<MutateProps<TData, TGraphQLVariables>>
>(
  document: DocumentNode,
  operationOptions: OperationOption<
    TProps,
    TData,
    TGraphQLVariables,
    TChildProps
  > = {}
): (
  WrappedComponent: React.ComponentType<TProps & TChildProps>
) => React.ComponentClass<TProps> {
  switch (parser(document).type) {
    case DocumentType.Mutation:
      return withMutation(document, operationOptions);
    case DocumentType.Subscription:
      return withSubscription(document, operationOptions);
    case DocumentType.Query:
    default:
      return withQuery(document, operationOptions);
  }
}
