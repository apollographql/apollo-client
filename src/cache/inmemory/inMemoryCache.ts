// Make builtins like Map and Set safe to use with non-extensible objects.
import './fixPolyfills';

import { DocumentNode } from 'graphql';
import { dep, wrap } from 'optimism';

import { ApolloCache } from '../core/cache';
import { Cache } from '../core/types/Cache';
import {
  addTypenameToDocument,
  StoreObject,
  Reference,
  isReference,
} from '../../utilities';
import {
  ApolloReducerConfig,
  NormalizedCacheObject,
} from './types';
import { StoreReader } from './readFromStore';
import { StoreWriter } from './writeToStore';
import { EntityStore, supportsResultCaching } from './entityStore';
import { makeVar } from './reactiveVars';
import {
  defaultDataIdFromObject,
  PossibleTypesMap,
  Policies,
  TypePolicies,
} from './policies';
import { hasOwn } from './helpers';

export interface InMemoryCacheConfig extends ApolloReducerConfig {
  resultCaching?: boolean;
  possibleTypes?: PossibleTypesMap;
  typePolicies?: TypePolicies;
}

const defaultConfig: InMemoryCacheConfig = {
  dataIdFromObject: defaultDataIdFromObject,
  addTypename: true,
  resultCaching: true,
  typePolicies: {},
};

export class InMemoryCache extends ApolloCache<NormalizedCacheObject> {
  private data: EntityStore;
  private optimisticData: EntityStore;

  protected config: InMemoryCacheConfig;
  private watches = new Set<Cache.WatchOptions>();
  private addTypename: boolean;

  private typenameDocumentCache = new Map<DocumentNode, DocumentNode>();
  private storeReader: StoreReader;
  private storeWriter: StoreWriter;

  // Dynamically imported code can augment existing typePolicies or
  // possibleTypes by calling cache.policies.addTypePolicies or
  // cache.policies.addPossibletypes.
  public readonly policies: Policies;

  public readonly makeVar = makeVar;

  constructor(config: InMemoryCacheConfig = {}) {
    super();
    this.config = { ...defaultConfig, ...config };
    this.addTypename = !!this.config.addTypename;

    this.policies = new Policies({
      cache: this,
      dataIdFromObject: this.config.dataIdFromObject,
      possibleTypes: this.config.possibleTypes,
      typePolicies: this.config.typePolicies,
    });

    // Passing { resultCaching: false } in the InMemoryCache constructor options
    // will completely disable dependency tracking, which will improve memory
    // usage but worsen the performance of repeated reads.
    this.data = new EntityStore.Root({
      policies: this.policies,
      resultCaching: this.config.resultCaching,
    });

    // When no optimistic writes are currently active, cache.optimisticData ===
    // cache.data, so there are no additional layers on top of the actual data.
    // When an optimistic update happens, this.optimisticData will become a
    // linked list of OptimisticCacheLayer objects that terminates with the
    // original this.data cache object.
    this.optimisticData = this.data;

    this.storeWriter = new StoreWriter(
      this,
      this.storeReader = new StoreReader({
        cache: this,
        addTypename: this.addTypename,
      }),
    );
  }

  public restore(data: NormalizedCacheObject): this {
    if (data) this.data.replace(data);
    return this;
  }

  public extract(optimistic: boolean = false): NormalizedCacheObject {
    return (optimistic ? this.optimisticData : this.data).toObject();
  }

  public read<T>(options: Cache.ReadOptions): T | null {
    const store = options.optimistic ? this.optimisticData : this.data;
    if (typeof options.rootId === 'string' && !store.has(options.rootId)) {
      return null;
    }
    return this.storeReader.diffQueryAgainstStore<T>({
      store,
      query: options.query,
      variables: options.variables,
      rootId: options.rootId,
      config: this.config,
      returnPartialData: false,
    }).result || null;
  }

  public write(options: Cache.WriteOptions): Reference | undefined {
    try {
      ++this.txCount;
      return this.storeWriter.writeToStore({
        store: this.data,
        query: options.query,
        result: options.result,
        dataId: options.dataId,
        variables: options.variables,
      });
    } finally {
      if (!--this.txCount && options.broadcast !== false) {
        this.broadcastWatches();
      }
    }
  }

  public modify(options: Cache.ModifyOptions): boolean {
    if (hasOwn.call(options, "id") && !options.id) {
      // To my knowledge, TypeScript does not currently provide a way to
      // enforce that an optional property?:type must *not* be undefined
      // when present. That ability would be useful here, because we want
      // options.id to default to ROOT_QUERY only when no options.id was
      // provided. If the caller attempts to pass options.id with a
      // falsy/undefined value (perhaps because cache.identify failed), we
      // should not assume the goal was to modify the ROOT_QUERY object.
      // We could throw, but it seems natural to return false to indicate
      // that nothing was modified.
      return false;
    }
    const store = options.optimistic // Defaults to false.
      ? this.optimisticData
      : this.data;
    try {
      ++this.txCount;
      return store.modify(options.id || "ROOT_QUERY", options.fields);
    } finally {
      if (!--this.txCount && options.broadcast !== false) {
        this.broadcastWatches();
      }
    }
  }

  public diff<T>(options: Cache.DiffOptions): Cache.DiffResult<T> {
    return this.storeReader.diffQueryAgainstStore({
      store: options.optimistic ? this.optimisticData : this.data,
      rootId: options.id || "ROOT_QUERY",
      query: options.query,
      variables: options.variables,
      returnPartialData: options.returnPartialData,
      config: this.config,
    });
  }

  public watch(watch: Cache.WatchOptions): () => void {
    this.watches.add(watch);
    if (watch.immediate) {
      this.maybeBroadcastWatch(watch);
    }
    return () => {
      this.watches.delete(watch);
    };
  }

  // Request garbage collection of unreachable normalized entities.
  public gc() {
    return this.optimisticData.gc();
  }

  // Call this method to ensure the given root ID remains in the cache after
  // garbage collection, along with its transitive child entities. Note that
  // the cache automatically retains all directly written entities. By default,
  // the retainment persists after optimistic updates are removed. Pass true
  // for the optimistic argument if you would prefer for the retainment to be
  // discarded when the top-most optimistic layer is removed. Returns the
  // resulting (non-negative) retainment count.
  public retain(rootId: string, optimistic?: boolean): number {
    return (optimistic ? this.optimisticData : this.data).retain(rootId);
  }

  // Call this method to undo the effect of the retain method, above. Once the
  // retainment count falls to zero, the given ID will no longer be preserved
  // during garbage collection, though it may still be preserved by other safe
  // entities that refer to it. Returns the resulting (non-negative) retainment
  // count, in case that's useful.
  public release(rootId: string, optimistic?: boolean): number {
    return (optimistic ? this.optimisticData : this.data).release(rootId);
  }

  // Returns the canonical ID for a given StoreObject, obeying typePolicies
  // and keyFields (and dataIdFromObject, if you still use that). At minimum,
  // the object must contain a __typename and any primary key fields required
  // to identify entities of that type. If you pass a query result object, be
  // sure that none of the primary key fields have been renamed by aliasing.
  // If you pass a Reference object, its __ref ID string will be returned.
  public identify(object: StoreObject | Reference): string | undefined {
    return isReference(object) ? object.__ref :
      this.policies.identify(object)[0];
  }

  public evict(options: Cache.EvictOptions): boolean {
    if (!options.id) {
      if (hasOwn.call(options, "id")) {
        // See comment in modify method about why we return false when
        // options.id exists but is falsy/undefined.
        return false;
      }
      options = { ...options, id: "ROOT_QUERY" };
    }
    try {
      // It's unlikely that the eviction will end up invoking any other
      // cache update operations while it's running, but {in,de}crementing
      // this.txCount still seems like a good idea, for uniformity with
      // the other update methods.
      ++this.txCount;
      return this.optimisticData.evict(options);
    } finally {
      if (!--this.txCount && options.broadcast !== false) {
        this.broadcastWatches();
      }
    }
  }

  public reset(): Promise<void> {
    this.data.clear();
    this.optimisticData = this.data;
    this.broadcastWatches();
    return Promise.resolve();
  }

  public removeOptimistic(idToRemove: string) {
    const newOptimisticData = this.optimisticData.removeLayer(idToRemove);
    if (newOptimisticData !== this.optimisticData) {
      this.optimisticData = newOptimisticData;
      this.broadcastWatches();
    }
  }

  private txCount = 0;

  public performTransaction(
    transaction: (cache: InMemoryCache) => any,
    optimisticId?: string | null,
  ) {
    const perform = (layer?: EntityStore) => {
      const { data, optimisticData } = this;
      ++this.txCount;
      if (layer) {
        this.data = this.optimisticData = layer;
      }
      try {
        transaction(this);
      } finally {
        --this.txCount;
        this.data = data;
        this.optimisticData = optimisticData;
      }
    };

    let fromOptimisticTransaction = false;

    if (typeof optimisticId === 'string') {
      // Note that there can be multiple layers with the same optimisticId.
      // When removeOptimistic(id) is called for that id, all matching layers
      // will be removed, and the remaining layers will be reapplied.
      this.optimisticData = this.optimisticData.addLayer(optimisticId, perform);
      fromOptimisticTransaction = true;
    } else if (optimisticId === null) {
      // Ensure both this.data and this.optimisticData refer to the root
      // (non-optimistic) layer of the cache during the transaction. Note
      // that this.data could be a Layer if we are currently executing an
      // optimistic transaction function, but otherwise will always be an
      // EntityStore.Root instance.
      perform(this.data);
    } else {
      // Otherwise, leave this.data and this.optimisticData unchanged and
      // run the transaction with broadcast batching.
      perform();
    }

    // This broadcast does nothing if this.txCount > 0.
    this.broadcastWatches(fromOptimisticTransaction);
  }

  public transformDocument(document: DocumentNode): DocumentNode {
    if (this.addTypename) {
      let result = this.typenameDocumentCache.get(document);
      if (!result) {
        result = addTypenameToDocument(document);
        this.typenameDocumentCache.set(document, result);
        // If someone calls transformDocument and then mistakenly passes the
        // result back into an API that also calls transformDocument, make sure
        // we don't keep creating new query documents.
        this.typenameDocumentCache.set(result, result);
      }
      return result;
    }
    return document;
  }

  protected broadcastWatches(fromOptimisticTransaction?: boolean) {
    if (!this.txCount) {
      this.watches.forEach(c => this.maybeBroadcastWatch(c, fromOptimisticTransaction));
    }
  }

  private maybeBroadcastWatch = wrap((
    c: Cache.WatchOptions,
    fromOptimisticTransaction?: boolean,
  ) => {
    return this.broadcastWatch.call(this, c, !!fromOptimisticTransaction);
  }, {
    makeCacheKey: (c: Cache.WatchOptions) => {
      // Return a cache key (thus enabling result caching) only if we're
      // currently using a data store that can track cache dependencies.
      const store = c.optimistic ? this.optimisticData : this.data;
      if (supportsResultCaching(store)) {
        const { optimistic, rootId, variables } = c;
        return store.makeCacheKey(
          c.query,
          // Different watches can have the same query, optimistic
          // status, rootId, and variables, but if their callbacks are
          // different, the (identical) result needs to be delivered to
          // each distinct callback. The easiest way to achieve that
          // separation is to include c.callback in the cache key for
          // maybeBroadcastWatch calls. See issue #5733.
          c.callback,
          JSON.stringify({ optimistic, rootId, variables }),
        );
      }
    }
  });

  private watchDep = dep<Cache.WatchOptions>();

  // This method is wrapped by maybeBroadcastWatch, which is called by
  // broadcastWatches, so that we compute and broadcast results only when
  // the data that would be broadcast might have changed. It would be
  // simpler to check for changes after recomputing a result but before
  // broadcasting it, but this wrapping approach allows us to skip both
  // the recomputation and the broadcast, in most cases.
  private broadcastWatch(
    c: Cache.WatchOptions,
    fromOptimisticTransaction: boolean,
  ) {
    // First, invalidate any other maybeBroadcastWatch wrapper functions
    // currently depending on this Cache.WatchOptions object (including
    // the one currently calling broadcastWatch), so they will be included
    // in the next broadcast, even if the result they receive is the same
    // as the previous result they received. This is important because we
    // are about to deliver a different result to c.callback, so any
    // previous results should have a chance to be redelivered.
    this.watchDep.dirty(c);

    // Next, re-depend on this.watchDep for just this invocation of
    // maybeBroadcastWatch (this is a no-op if broadcastWatch was not
    // called by maybeBroadcastWatch). This allows only the most recent
    // maybeBroadcastWatch invocation for this watcher to remain cached,
    // enabling re-broadcast of previous results even if they have not
    // changed since they were previously delivered.
    this.watchDep(c);

    const diff = this.diff<any>({
      query: c.query,
      variables: c.variables,
      optimistic: c.optimistic,
    });

    if (c.optimistic && fromOptimisticTransaction) {
      diff.fromOptimisticTransaction = true;
    }

    c.callback(diff);
  }
}
