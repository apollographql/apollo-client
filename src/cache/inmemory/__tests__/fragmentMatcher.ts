import gql from 'graphql-tag';

import { InMemoryCache } from '../inMemoryCache';
import { visit, FragmentDefinitionNode } from 'graphql';
import { hasOwn } from '../helpers';

describe('fragment matching', () => {
  it('can match exact types with or without possibleTypes', () => {
    const cacheWithoutPossibleTypes = new InMemoryCache({
      addTypename: true,
    });

    const cacheWithPossibleTypes = new InMemoryCache({
      addTypename: true,
      possibleTypes: {
        Animal: ['Cat', 'Dog'],
      },
    });

    const query = gql`
      query AnimalNames {
        animals {
          id
          name
          ...CatDetails
        }
      }
      fragment CatDetails on Cat {
        livesLeft
        killsToday
      }
    `;

    const data = {
      animals: [
        {
          __typename: 'Cat',
          id: 1,
          name: 'Felix',
          livesLeft: 8,
          killsToday: 2,
        },
        {
          __typename: 'Dog',
          id: 2,
          name: 'Baxter',
        },
      ],
    };

    cacheWithoutPossibleTypes.writeQuery({ query, data });
    expect(cacheWithoutPossibleTypes.readQuery({ query })).toEqual(data);

    cacheWithPossibleTypes.writeQuery({ query, data });
    expect(cacheWithPossibleTypes.readQuery({ query })).toEqual(data);
  });

  it('can match interface subtypes', () => {
    const cache = new InMemoryCache({
      addTypename: true,
      possibleTypes: {
        Animal: ['Cat', 'Dog'],
      },
    });

    const query = gql`
      query BestFriend {
        bestFriend {
          id
          ...AnimalName
        }
      }
      fragment AnimalName on Animal {
        name
      }
    `;

    const data = {
      bestFriend: {
        __typename: 'Dog',
        id: 2,
        name: 'Beckett',
      },
    };

    cache.writeQuery({ query, data });
    expect(cache.readQuery({ query })).toEqual(data);
  });

  it('can match union member types', () => {
    const cache = new InMemoryCache({
      addTypename: true,
      possibleTypes: {
        Status: ['PASSING', 'FAILING', 'SKIPPED'],
      },
    });

    const query = gql`
      query {
        testResults {
          id
          output {
            ... on Status {
              stdout
            }
            ... on FAILING {
              stderr
            }
          }
        }
      }
    `;

    const data = {
      testResults: [
        {
          __typename: 'TestResult',
          id: 123,
          output: {
            __typename: 'PASSING',
            stdout: 'ok!',
          },
        },
        {
          __typename: 'TestResult',
          id: 456,
          output: {
            __typename: 'FAILING',
            stdout: '',
            stderr: 'oh no',
          },
        },
      ],
    };

    cache.writeQuery({ query, data });
    expect(cache.readQuery({ query })).toEqual(data);
  });

  it('can match indirect subtypes while avoiding cycles', () => {
    const cache = new InMemoryCache({
      addTypename: true,
      possibleTypes: {
        Animal: ['Animal', 'Bug', 'Mammal'],
        Bug: ['Ant', 'Spider', 'RolyPoly'],
        Mammal: ['Dog', 'Cat', 'Human'],
        Cat: ['Calico', 'Siamese', 'Sphynx', 'Tabby'],
      },
    });

    const query = gql`
      query {
        animals {
          ... on Mammal {
            hasFur
            bodyTemperature
          }
          ... on Bug {
            isVenomous
          }
        }
      }
    `;

    const data = {
      animals: [
        {
          __typename: 'Sphynx',
          hasFur: false,
          bodyTemperature: 99,
        },
        {
          __typename: 'Dog',
          hasFur: true,
          bodyTemperature: 102,
        },
        {
          __typename: 'Spider',
          isVenomous: 'maybe',
        },
      ],
    };

    cache.writeQuery({ query, data });
    expect(cache.readQuery({ query })).toEqual(data);
  });

  it('can match against the root Query', () => {
    const cache = new InMemoryCache({
      addTypename: true,
    });

    const query = gql`
      query AllPeople {
        people {
          id
          name
        }
        ...PeopleTypes
      }
      fragment PeopleTypes on Query {
        __type(name: "Person") {
          name
          kind
        }
      }
    `;

    const data = {
      people: [
        {
          __typename: 'Person',
          id: 123,
          name: 'Ben',
        },
      ],
      __type: {
        __typename: '__Type',
        name: 'Person',
        kind: 'OBJECT',
      },
    };

    cache.writeQuery({ query, data });
    expect(cache.readQuery({ query })).toEqual(data);
  });

});

describe("policies.fragmentMatches", () => {
  const warnings: any[] = [];
  const { warn } = console;

  beforeEach(() => {
    warnings.length = 0;
    console.warn = function (message: any) {
      warnings.push(message);
    };
  });

  afterEach(() => {
    console.warn = warn;
  });

  it("can infer fuzzy subtypes heuristically", () => {
    const cache = new InMemoryCache({
      possibleTypes: {
        A: ["B", "C"],
        B: ["D"],
        C: ["[E-Z]"],
      },
    });

    const fragments = gql`
      fragment FragA on A { a }
      fragment FragB on B { b }
      fragment FragC on C { c }
      fragment FragD on D { d }
      fragment FragE on E { e }
      fragment FragF on F { f }
    `;

    function checkTypes(
      expected: Record<string, Record<string, boolean>>,
    ) {
      const checked = new Set<FragmentDefinitionNode>();

      visit(fragments, {
        FragmentDefinition(frag) {
          function check(typename: string, result: boolean) {
            if (result !== cache.policies.fragmentMatches(frag, typename)) {
              fail(`fragment ${
                frag.name.value
              } should${result ? "" : " not"} have matched typename ${typename}`);
            }
          }

          const supertype = frag.typeCondition.name.value;
          expect("ABCDEF".split("")).toContain(supertype);

          if (hasOwn.call(expected, supertype)) {
            Object.keys(expected[supertype]).forEach(subtype => {
              check(subtype, expected[supertype][subtype]);
            });

            checked.add(frag);
          }
        },
      });

      return checked;
    }

    expect(checkTypes({
      A: {
        A: true,
        B: true,
        C: true,
        D: true,
        E: false,
        F: false,
        G: false,
      },
      B: {
        A: false,
        B: true,
        C: false,
        D: true,
        E: false,
        F: false,
        G: false,
      },
      C: {
        A: false,
        B: false,
        C: true,
        D: false,
        E: false,
        F: false,
        G: false,
      },
      D: {
        A: false,
        B: false,
        C: false,
        D: true,
        E: false,
        F: false,
        G: false,
      },
      E: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: true,
        F: false,
        G: false,
      },
      F: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: true,
        G: false,
      },
      G: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: false,
        G: true,
      },
    }).size).toBe("ABCDEF".length);

    cache.writeQuery({
      query: gql`
        query {
          objects {
            ...FragC
          }
        }
        ${fragments}
      `,
      data: {
        objects: [
          { __typename: "E", c: "ce" },
          { __typename: "F", c: "cf" },
          { __typename: "G", c: "cg" },
          // The /[E-Z]/ subtype pattern specified for the C supertype
          // must match the entire subtype string.
          { __typename: "TooLong", c: "nope" },
          // The H typename matches the regular expression for C, but it
          // does not pass the heuristic test of having all the fields
          // expected if FragC matched.
          { __typename: "H", h: "not c" },
        ],
      },
    });

    expect(warnings).toEqual([
      "Inferring subtype E of supertype C",
      "Inferring subtype F of supertype C",
      "Inferring subtype G of supertype C",
      // Note that TooLong is not inferred here.
    ]);

    expect(checkTypes({
      A: {
        A: true,
        B: true,
        C: true,
        D: true,
        E: true,
        F: true,
        G: true,
        H: false,
      },
      B: {
        A: false,
        B: true,
        C: false,
        D: true,
        E: false,
        F: false,
        G: false,
        H: false,
      },
      C: {
        A: false,
        B: false,
        C: true,
        D: false,
        E: true,
        F: true,
        G: true,
        H: false,
      },
      D: {
        A: false,
        B: false,
        C: false,
        D: true,
        E: false,
        F: false,
        G: false,
        H: false,
      },
      E: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: true,
        F: false,
        G: false,
        H: false,
      },
      F: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: true,
        G: false,
        H: false,
      },
      G: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: true,
        G: true,
        H: false,
      },
    }).size).toBe("ABCDEF".length);

    expect(cache.extract()).toMatchSnapshot();

    // Now add the TooLong subtype of C explicitly.
    cache.policies.addPossibleTypes({
      C: ["TooLong"],
    });

    expect(checkTypes({
      A: {
        A: true,
        B: true,
        C: true,
        D: true,
        E: true,
        F: true,
        G: true,
        TooLong: true,
        H: false,
      },
      B: {
        A: false,
        B: true,
        C: false,
        D: true,
        E: false,
        F: false,
        G: false,
        TooLong: false,
        H: false,
      },
      C: {
        A: false,
        B: false,
        C: true,
        D: false,
        E: true,
        F: true,
        G: true,
        TooLong: true,
        H: false,
      },
      D: {
        A: false,
        B: false,
        C: false,
        D: true,
        E: false,
        F: false,
        G: false,
        TooLong: false,
        H: false,
      },
      E: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: true,
        F: false,
        G: false,
        TooLong: false,
        H: false,
      },
      F: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: true,
        G: false,
        TooLong: false,
        H: false,
      },
      G: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: true,
        G: true,
        TooLong: false,
        H: false,
      },
      H: {
        A: false,
        B: false,
        C: false,
        D: false,
        E: false,
        F: false,
        G: false,
        TooLong: false,
        H: true,
      },
    }).size).toBe("ABCDEF".length);
  });
});
