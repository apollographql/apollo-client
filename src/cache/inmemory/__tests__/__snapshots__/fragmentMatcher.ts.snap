// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`policies.fragmentMatches can infer fuzzy subtypes heuristically 1`] = `
Object {
  "ROOT_QUERY": Object {
    "__typename": "Query",
    "objects": Array [
      Object {
        "__typename": "E",
        "c": "ce",
      },
      Object {
        "__typename": "F",
        "c": "cf",
      },
      Object {
        "__typename": "G",
        "c": "cg",
      },
      Object {
        "__typename": "TooLong",
      },
      Object {
        "__typename": "H",
      },
    ],
  },
}
`;
